idaes/apps/grid_integration/pricetaker/clustering.py

```diff
@@ -181,28 +193,93 @@ def get_optimal_num_clusters(
 
     k_values = list(range(kmin, kmax + 1))
     inertia_values = []
-    mean_silhouette = []
 
     for k in k_values:
         kmeans = KMeans(n_clusters=k, random_state=seed).fit(samples)
         inertia_values.append(kmeans.inertia_)
 
-        if method == "silhouette":
-            # Calculate the average silhouette score, if the chosen method
-            # is silhouette
-            mean_silhouette.append(silhouette_score(samples, kmeans.labels_))
-
-    # Identify the optimal number of clusters
-    if method == "elbow":
-        raise NotImplementedError(
-            "elbow method is not supported currently for finding the optimal "
-            "number of clusters."
-        )
+    if method == "silhouette":
+        mean_silhouette = []
+        mean_silhouette.append(silhouette_score(samples, kmeans.labels_))
 
-    elif method == "silhouette":
+        # Identify the optimal number of clusters
         max_index = mean_silhouette.index(max(mean_silhouette))
         n_clusters = k_values[max_index]
 
+    # The implementation below is based on
+    # Ville Satopaa, Jeannie Albrecht, David Irwin, Barath Raghavan
+    # Finding a “Kneedle” in a Haystack:
+    # Detecting Knee Points in System Behavior
+    # https://raghavan.usc.edu/papers/kneedle-simplex11.pdf
+
+    elif method == "elbow":
+        # Invert inertia values such that plot is concave down and increasing
+        inverted_inertia_values = [-y for y in inertia_values]
+        # Use a smoothing spline that retains the data's original shape
+        spline = splrep(k_values, inverted_inertia_values, k=3)
+        k_smooth = np.linspace(kmin, kmax + 1, 100)
+        inertia_smooth = splev(k_smooth, spline)
+
+        k_norm = _normalize_values(k_smooth)
+        inertia_norm = _normalize_values(inertia_smooth)
+
+        # Compute the set of differences (x, y) -> (x, y-x)
+        inertia_diff = []
+        for i in range(len(inertia_norm)):
+            set_of_differences = inertia_norm[i] - k_norm[i]
+            inertia_diff.append(set_of_differences)
+
+        # Identify local maxima
+        local_maxima = [
+            (k_norm[i], inertia_diff[i])
+            for i in range(1, len(inertia_diff) - 1)
+            if inertia_diff[i] > inertia_diff[i - 1]
+            and inertia_diff[i] > inertia_diff[i + 1]
+        ]
+
+        # Calculate optimal # of clusters based on the # of local maxima
+        threshold_values = []
+        threshold_triggered = False
+        n_clusters = 0
+
+        if len(local_maxima) == 0:
+            n_clusters = 0
+            _logger.warning(
```

Comment:
Instead of warning, raise an error: Optimal number of clusters cannot be determined with the elbow method for this dataset. Try silhouette method instead.